src/background/quality/providers/QualityProvider.test.ts

```diff
@@ -0,0 +1,144 @@
+});
+
+test("throws StorageError, if storage get operation fails", async () => {
+  mockBrowser.storage.local.get.expect(["qualities"]).andReject(new Error());
+
+  await expect(
+    qualityProvider["getQuality"]("https://www.example.com")
+  ).rejects.toBeInstanceOf(StorageError);
+});
+
+test("sets quality for provided url, if quality is stored already", async () => {
+  mockBrowser.storage.local.get.expect(["qualities"]).andResolve({
+    qualities: { "www.example.com": Quality.OrganizationValidated },
+  });
+  mockBrowser.storage.local.set
+    .expect({ qualities: { "www.example.com": Quality.DomainValidated } })
+    .andResolve();
+
+  await expect(
+    qualityProvider.setQuality(
+      "https://www.example.com",
+      Quality.DomainValidated
+    )
+  ).resolves.toEqual(undefined);
+});
+
+test("sets quality for provided url, if quality is not stored already", async () => {
+  mockBrowser.storage.local.get.expect(["qualities"]).andResolve({});
+  mockBrowser.storage.local.set
+    .expect({ qualities: { "www.example.com": Quality.DomainValidated } })
+    .andResolve();
+
+  await expect(
+    qualityProvider.setQuality(
+      "https://www.example.com",
+      Quality.DomainValidated
+    )
+  ).resolves.toEqual(undefined);
+});
+
+test("throws Storage, if storage set operation fails", async () => {
+  mockBrowser.storage.local.get.expect(["qualities"]).andResolve({});
+  mockBrowser.storage.local.set
+    .expect({ qualities: { "www.example.com": Quality.DomainValidated } })
+    .andReject(new Error());
+
+  await expect(
+    qualityProvider.setQuality(
+      "https://www.example.com",
+      Quality.DomainValidated
+    )
+  ).rejects.toBeInstanceOf(StorageError);
+});
+
+test("returns true, if provided quality is lower than stored quality", async () => {
+  mockBrowser.storage.local.get.expect(["qualities"]).andResolve({
+    qualities: { "www.example.com": Quality.ExtendedValidated },
+  });
+
+  await expect(
+    qualityProvider.hasQualityDecreased(
+      "https://www.example.com",
+      Quality.DomainValidated
+    )
+  ).resolves.toEqual(true);
+});
+
+test("returns false, if provided quality is higher than stored quality", async () => {
+  mockBrowser.storage.local.get.expect(["qualities"]).andResolve({
+    qualities: { "www.example.com": Quality.DomainValidated },
+  });
+
+  await expect(
+    qualityProvider.hasQualityDecreased(
+      "https://www.example.com",
+      Quality.OrganizationValidated
+    )
+  ).resolves.toEqual(false);
+});
+
+test("returns false, if provided quality is equal with stored quality", async () => {
+  mockBrowser.storage.local.get.expect(["qualities"]).andResolve({
+    qualities: { "www.example.com": Quality.DomainValidated },
+  });
+
+  await expect(
+    qualityProvider.hasQualityDecreased(
+      "https://www.example.com",
+      Quality.DomainValidated
+    )
+  ).resolves.toEqual(false);
+});
+
+test("sets quality of provided url to unknown", async () => {

```

Comment:
When?

Reply:
on reset?

Reply:
Yes, I've modified the text a bit to reflect that.